llvm/lib/Target/AMDGPU/AMDGPUAsmPrinter.h

```diff
@@ -40,12 +42,19 @@ class AMDGPUAsmPrinter final : public AsmPrinter {
 
   AMDGPUResourceUsageAnalysis *ResourceUsage;
 
+  MCResourceInfo RI;
+
   SIProgramInfo CurrentProgramInfo;
 
   std::unique_ptr<AMDGPU::HSAMD::MetadataStreamer> HSAMetadataStream;
 
   MCCodeEmitter *DumpCodeInstEmitter = nullptr;
 
+  // validateMCResourceInfo cannot recompute parts of the occupancy as it does
+  // for other metadata to validate (e.g., NumSGPRs) so a map is necessary if we
+  // really want to track and validate the occupancy.
```

Comment:
I don't see why this is the case

Reply:
Yeah, I thought I needed access to the `MachineFunction` for `SIMachineFunctionInfo` but I can reconstruct it using just the `Function` and `GCNSubTarget`. It now recomputes the occupancy instead of this caching.

Reply:
You do need the MachineFunction to get its SIMachineFunctionInfo. Constructing a new one won't give the same result?